src/ice_nucleation.F90

```diff
@@ -0,0 +1,563 @@
+          n_parts_in_bin = integer_varray_n_entry(aero_state%aero_sorted%size_class%inverse(i_bin, i_class))
+          radius_max = aero_state%aero_sorted%bin_grid%edges(i_bin + 1)
+          diameter_max = radius_max * 2
+          if (immersion_freezing_scheme_type .eq. &
+               IMMERSION_FREEZING_SCHEME_ABIFM) then
+             call ABIFM_max(diameter_max, p_freeze_max, aero_data, j_het_max, del_t)
+          else if (immersion_freezing_scheme_type .eq. &
+               IMMERSION_FREEZING_SCHEME_CONST) then
+             p_freeze_max = 1 - exp(freezing_rate * del_t)
+          endif
+
+          k_th = n_parts_in_bin + 1
+          loop_choosed_particles: do while(.TRUE.)
+             rand_geometric = pmc_random_geometric(p_freeze_max)
+             k_th = k_th - rand_geometric
+             if (k_th <= 0) then
+                EXIT loop_choosed_particles
+             endif
+             loop_count = loop_count + 1
+             i_part = aero_state%aero_sorted%size_class%inverse(i_bin, i_class)%entry(k_th)
+             if (aero_state%apa%particle(i_part)%frozen) then
+                cycle
+             end if
+             if (H2O_frac(i_part) < 1e-2) then
+                cycle
+             end if
+             if (immersion_freezing_scheme_type .eq. &
+                  IMMERSION_FREEZING_SCHEME_ABIFM) then
+                call ABIFM_particle(i_part, p_freeze, aero_state, aero_data, a_w_ice, del_t)
+                call warn_assert_msg(301184565, p_freeze .le. p_freeze_max,&
+                     "p_freeze > p_freeze_max.")
+                rand = pmc_random()
+                if (rand < p_freeze / p_freeze_max) then
+                   aero_state%apa%particle(i_part)%frozen = .TRUE.
+                   aero_state%apa%particle(i_part)%den_ice = &
+                        const%reference_ice_density
+                   aero_state%apa%particle(i_part)%ice_shape_phi = 1d0
+
+                endif 
+             else
+                aero_state%apa%particle(i_part)%frozen = .TRUE.
+                aero_state%apa%particle(i_part)%den_ice = &
+                     const%reference_ice_density
+                aero_state%apa%particle(i_part)%ice_shape_phi = 1d0
+             endif
+
+          enddo loop_choosed_particles
+       enddo loop_classes
+    enddo loop_bins
+    
+    deallocate(total_masses)
+    deallocate(H2O_masses)
+    deallocate(H2O_frac)
+     
+  end subroutine immersion_freezing_time_dependent
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Simulation for time-dependent scheme (e.g., ABIFM, constant rate),
+  !> deciding whether to freeze for each particle. Run in each time step.   
+  !> This subroutine applys the naive algorithm for reference.
+  subroutine immersion_freezing_time_dependent_naive(aero_state, aero_data, &
+    env_state_initial, env_state_final, del_t, &
+    immersion_freezing_scheme_type, &
+    freezing_rate)
```

Comment:
`env_state_initial` is passed here and never used (and if this is correct, `env_state_final` can just become `env_state`).